### PR TITLE
[TrimmableTypeMap] Add UCO wrappers, JCW generator, and JNI signature helpers

### LINKED ISSUE #10799: [TrimmableTypeMap] TypeMap assembly + JCW Java source generators
Part of https://github.com/dotnet/android/issues/10789
Depends on: Scanner sub-issue (PR1)

## Overview

Implement the IL assembly generator and JCW Java source generator that transform `JavaPeerInfo` scan results into:
1. **Per-assembly typemap assemblies** + root `_Microsoft.Android.TypeMaps.dll` — using SRM Ecma335 APIs
2. **JCW Java source files** — with `registerNatives` in `static {}` blocks

## TypeMap Assembly Generator (`TypeMapAssemblyGenerator`)

Uses SRM types: `MetadataBuilder`, `BlobBuilder`, `InstructionEncoder`, `ControlFlowBuilder`, `MethodBodyStreamEncoder`, `PEHeaderBuilder`.

### Generated assembly contents

**Assembly-level attributes:**
- `[assembly: IgnoresAccessChecksTo("Mono.Android")]` etc. for all referenced assemblies
- `[assembly: TypeMap<Java.Lang.Object>("jni/name", typeof(Proxy), typeof(TrimTarget))]` — one per type
- `[assembly: TypeMapAssociation<AliasesUniverse>(typeof(Source), typeof(AliasHolder))]` — for aliases

**Per-type proxy classes** — `JavaPeerProxyAttribute` subclasses:
- Self-application: `[Proxy] class Proxy : JavaPeerProxyAttribute`
- `CreateInstance(IntPtr, JniHandleOwnership)`:
  - MCW with own ctor: `new T(handle, transfer)`
  - User type without own ctor: `GetUninitializedObject(T)` + direct call to base ctor (via `IgnoresAccessChecksTo`)
  - Interface: `new TInvoker(handle, transfer)`
  - Generic: `throw new NotSupportedException()`
- `GetContainerFactory()`: `return JavaPeerContainerFactory.Create<T>()`
- For ACW types: `IAndroidCallableWrapper.RegisterNatives(JniType)` + `GetFunctionPointer(int)`
- `[UnmanagedCallersOnly]` UCO methods (delegation or inlined)

**Per-assembly output model** (for incremental builds):
- One typemap assembly per input assembly with Java peer types
- Root assembly uses `[assembly: TypeMapAssemblyTarget<JLO>(assemblyName)]` to reference per-assembly typemaps
- Alias detection and cross-ref forcing handled in root assembly

## JCW Java Source Generator (`JcwJavaSourceGenerator`)

Generates `.java` files for ACW types (`DoNotGenerateAcw=false`):

```java
package example;
public class MainActivity extends android.app.Activity
    implements mono.android.IGCUserPeer
{
    static { mono.android.Runtime.registerNatives(MainActivity.class); }

    public MainActivity() { super(); if (getClass() == MainActivity.class) nctor_0(); }
    private native void nctor_0();

    @Override
    public void onCreate(android.os.Bundle p0) { n_onCreate(p0); }
    public native void n_onCreate(android.os.Bundle p0);
}
```

Key: generate Java class and IL proxy from the same ordered method list to ensure `RegisterNatives` connects the right `native` methods to the right UCO methods.

## Tests

- Produces valid PE assembly loadable by `MetadataReader`
- Generated assembly has correct `TypeMapAttribute` entries (2-arg and 3-arg)
- Proxy types have correct self-application pattern
- `CreateInstance` IL is valid for each ctor style
- `RegisterNatives` method has correct JNI registrations
- UCO wrappers have `[UnmanagedCallersOnly]` attribute
- Non-blittable params converted correctly (`byte` for `bool`)
- Java source has `registerNatives` in `static {}` block
- Correct `nctor_N` declarations and `@Override` annotations
- Package names from `[Register]`, not CRC-hashed

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes https://github.com/dotnet/android/issues/10799

## Summary

Adds UCO (UnifiedCallableObject) wrapper generation, JCW (Java Callable Wrapper) Java source generation, and JNI signature parsing utilities on top of the TypeMap proxy groundwork from PR #10808.

### Generator additions

- **JcwJavaSourceGenerator** — generates `.java` source files for each ACW type with `registerNatives` in `static {}` blocks, `native` method declarations, constructor chains with `super()` calls, and `@Override` annotations
- **JniSignatureHelper** — JNI signature parsing (parameter types, return types, JNI↔Java type conversion, CLR type encoding for IL emission)
- **ModelBuilder** — extended with UCO method/constructor building, native registration data, and `isAcw` classification
- **TypeMapAssemblyEmitter** — extended with `EmitUcoMethod` (UnmanagedCallersOnly wrappers that delegate to n_* callbacks), `EmitUcoConstructor`, `EmitRegisterNatives` (JNI native method registration), and `IAndroidCallableWrapper` interface implementation for ACW proxy types

### Scanner enrichment

- **JavaPeerInfo** — added `BaseJavaName`, `ImplementedInterfaceJavaNames`, `JavaConstructors`, `JniParameterInfo`, `ThrownNames` for downstream generators
- **JavaPeerScanner** — resolver methods for base types, interfaces, constructors; fixed `NativeCallbackName` for constructors (`n_ctor` instead of invalid `n_.ctor`)

### Model additions

- **UcoMethodData** / **UcoConstructorData** — describe UnmanagedCallersOnly wrappers to emit
- **NativeRegistrationData** — JNI native method registrations for RegisterNatives

### Tests

- 215 tests covering all generators, model builder scenarios, scanner behavior, and edge cases
- Test fixtures extended with `ExportWithThrows` type
- Stream-based test helpers avoid disk I/O for faster, more reliable tests

## Follow-up work

- **`GetContainerFactory()`** — moved to #10791 (runtime issue); will emit the method body once `JavaPeerContainerFactory` is defined
- **MSBuild task integration** (`_GenerateJavaStubs` decomposition) — tracked in #10807
- **Implementor/EventDispatcher trimming optimization** — tracked in #10911

## Note on `GetFunctionPointer`

The issue spec mentioned `GetFunctionPointer(int)` as an alternative dispatch mechanism. Since we use `RegisterNatives` (which binds JNI native methods to UCO wrappers at class load time), `GetFunctionPointer` is not needed.

### CODE REVIEW

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 1]
```diff
@@ -550,17 +498,114 @@ void EmitTypeGetter (string methodName, TypeRefData typeRef, MethodAttributes at
...
 		var ctorRef = entry.IsUnconditional ? _typeMapAttrCtorRef2Arg : _typeMapAttrCtorRef3Arg;
 		var blob = _pe.BuildAttributeBlob (b => {
 			b.WriteSerializedString (entry.JniName);
 			b.WriteSerializedString (entry.ProxyTypeReference);
 			if (!entry.IsUnconditional) {
-				if (entry.TargetTypeReference is null) {
-					throw new InvalidOperationException ($"TargetTypeReference must not be null for conditional entry '{entry.JniName}'");
-				}
-				b.WriteSerializedString (entry.TargetTypeReference);
+				b.WriteSerializedString (entry.TargetTypeReference!);
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs]: 🤖 ❌ **Nullable** — The null-forgiving operator `!` is used here on `entry.TargetTypeReference!`. This is banned in this repo. Since the guard `!entry.IsUnconditional` guarantees `TargetTypeReference != null` (because `IsUnconditional => TargetTypeReference == null`), assign to a local with an explicit null check:

```csharp
var targetRef = entry.TargetTypeReference ?? throw new InvalidOperationException ("TargetTypeReference must be set for conditional entries");
b.WriteSerializedString (targetRef);
```

_Rule: Never use `!` (null-forgiving operator)_

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ⚠️ Needs Changes

Found **4 issues**: 1 error, 1 warning, 2 suggestions.

- ❌ **Nullable** — Null-forgiving operator `!` used on `TargetTypeReference` (`TypeMapAssemblyEmitter.cs:608`)
- ⚠️ **Error handling** — `ParseSingleType` doesn't validate `IndexOf(';')` result for malformed object types (`JniSignatureHelper.cs:93`)
- 💡 **Formatting** — Use `[]` instead of `Array.Empty<T>()` per repo convention (multiple files)
- 💡 **Code organization** — Missing `#nullable enable` on new files

👍 Clean separation of scanner → model → emitter pipeline. Good use of `sealed record` for immutable data. `EmitCreateInstance` is well-decomposed into named methods. Stream overloads for testability are a nice touch.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/JniSignatureHelper.cs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+		case 'Z': i++; return JniParamKind.Boolean;
+		case 'B': i++; return JniParamKind.Byte;
+		case 'C': i++; return JniParamKind.Char;
+		case 'S': i++; return JniParamKind.Short;
+		case 'I': i++; return JniParamKind.Int;
+		case 'J': i++; return JniParamKind.Long;
+		case 'F': i++; return JniParamKind.Float;
+		case 'D': i++; return JniParamKind.Double;
+		case 'L':
+			i = sig.IndexOf (';', i) + 1;
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/JniSignatureHelper.cs]: 🤖 ⚠️ **Error handling** — If a malformed JNI signature has an `L` prefix without a closing `;`, `IndexOf(';', i)` returns `-1`, making `i = 0`. This silently restarts parsing from the beginning, causing incorrect results or an infinite loop in the caller's `while` loop. Consider validating:

```csharp
case 'L':
    int semi = sig.IndexOf (';', i);
    if (semi < 0)
        throw new ArgumentException ($"Malformed object type in '{sig}' at index {i}: missing ';'");
    i = semi + 1;
    return JniParamKind.Object;
```

Similar concern at lines 65 and 76 for `ParseReturnTypeString`/`ParseReturnType` where `IndexOf(')')` could return `-1`.

_Rule: Validate parameters — throw for unexpected values_

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerInfo.cs [hunk 3]
```diff
@@ -80,6 +80,12 @@ sealed record JavaPeerInfo
 	/// </summary>
 	public IReadOnlyList<MarshalMethodInfo> MarshalMethods { get; init; } = Array.Empty<MarshalMethodInfo> ();
 
+	/// <summary>
+	/// Java constructors to emit in the JCW .java file.
+	/// Each has a JNI signature and an ordinal index for the nctor_N native method.
+	/// </summary>
+	public IReadOnlyList<JavaConstructorInfo> JavaConstructors { get; init; } = Array.Empty<JavaConstructorInfo> ();
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Scanner/JavaPeerInfo.cs]: 🤖 💡 **Formatting** — Use `[]` instead of `Array.Empty<JavaConstructorInfo> ()` per repo convention. Same applies to `Parameters` properties at lines 160 and 220.

_Rule: Use `[]` not `Array.Empty<T>()`_

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/JcwJavaSourceGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,287 @@
+using System;
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/JcwJavaSourceGenerator.cs:1]: 🤖 💡 **Code organization** — New files should have `#nullable enable` at the top (no preceding blank lines). This applies to all new `.cs` files in this PR: `JcwJavaSourceGenerator.cs`, `JniSignatureHelper.cs`, `TypeMapAssemblyData.cs`. Note: this is a `netstandard2.0` project, so keep using `if (x is null) throw new ArgumentNullException (nameof (x))` instead of `ThrowIfNull`.

_Rule: New files should have `#nullable enable`_

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/ModelBuilder.cs [hunk 5]
```diff
@@ -50,7 +50,7 @@ public static TypeMapAssemblyData Build (IReadOnlyList<JavaPeerInfo> peers, stri
 		// Invoker types are NOT emitted as separate proxies or TypeMap entries —
 		// they only appear as a TypeRef in the interface proxy's get_InvokerType property.
 		var invokerTypeNames = new HashSet<string> (
-			peers.Select (p => p.InvokerTypeName).OfType<string> (),
+			peers.Where (p => p.InvokerTypeName != null).Select (p => p.InvokerTypeName!),
```
Comment (simonrozsival) [inline src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/ModelBuilder.cs]: 🤖 ❌ **Nullable** — `p.InvokerTypeName!` uses the banned null-forgiving operator. The `Where` clause guarantees non-null, but the compiler doesn't track that through the LINQ chain. Use the null-coalescing operator instead:

```csharp
peers.Where (p => p.InvokerTypeName != null).Select (p => p.InvokerTypeName ?? ""),
```

Or use a local function that narrows the type.

_Rule: Never use `!` (null-forgiving operator)_

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/TypeMapModelBuilderTests.cs [hunk 6]
```diff
@@ -303,21 +765,120 @@ public void Fixture_Service_NoActivation_NoProxy ()
...
+			Assert.True (proxy!.IsAcw);
+			Assert.True (proxy.HasActivation);
+		}
+
+		[Fact]
+		public void Fixture_MainActivity_UcoMethods ()
+		{
+			var peer = FindFixtureByJavaName ("my/app/MainActivity");
+			var model = BuildModel (new [] { peer }, "TypeMap");
+			var proxy = FindProxy (model, "MyApp_MainActivity_Proxy")!;
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/TypeMapModelBuilderTests.cs]: 🤖 ❌ **Nullable** — `FindProxy(model, "MyApp_MainActivity_Proxy")!` uses the banned null-forgiving operator. Replace with an explicit Assert + non-null variable:

```csharp
var proxy = FindProxy (model, "MyApp_MainActivity_Proxy");
Assert.NotNull (proxy);
```

`Assert.NotNull` already fails the test if null, and xUnit's NRT annotations make the variable non-null after the assertion.

_Rule: Never use `!` (null-forgiving operator)_

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs [hunk 7]
```diff
@@ -109,4 +126,17 @@ private protected static List<string> GetMemberRefNames (MetadataReader reader)
...
+	{
+		var dir = Path.Combine (Path.GetTempPath (), $"typemap-test-{Guid.NewGuid ():N}");
+		Directory.CreateDirectory (dir);
+		return dir;
+	}
+
+	private protected static void DeleteTempDir (string dir)
+	{
+		if (Directory.Exists (dir))
+			try { Directory.Delete (dir, true); } catch { }
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs]: 🤖 ⚠️ **Error handling** — Empty `catch { }` swallows all exceptions silently. For test cleanup, this is arguably acceptable (we don't want a failing `DeleteTempDir` to mask the real test failure), but best practice is to at least catch a specific exception type:

```csharp
try { Directory.Delete (dir, true); } catch (IOException) { }
```

This way unexpected exceptions (e.g., `UnauthorizedAccessException`) still surface.

_Rule: No empty catch blocks — challenge exception swallowing_

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ⚠️ Needs Changes

Found **3 issues**: 2 errors, 1 warning.

- ❌ **Nullable**: Null-forgiving `!` on `InvokerTypeName` (`ModelBuilder.cs:53`)
- ❌ **Nullable**: Null-forgiving `!` on `FindProxy` result (`TypeMapModelBuilderTests.cs:791`)
- ⚠️ **Error handling**: Empty catch block swallows all exceptions (`FixtureTestBase.cs:140`)

👍 Previous round of findings (IndexOf validation, `#nullable enable`, `Array.Empty` → `[]`) all addressed correctly. The JNI signature parsing is now robust against malformed inputs.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ✅ LGTM

Found **0 issues**.

All previously flagged concerns have been addressed:
- ✅ No null-forgiving `!` operators remaining
- ✅ `IndexOf` results validated in `JniSignatureHelper` (throws on malformed input)
- ✅ `#nullable enable` on all new files
- ✅ `Array.Empty<T>()` → `[]` in new code
- ✅ Empty `catch { }` narrowed to `catch (IOException) { }`

👍 Clean pipeline architecture (scanner → model → emitter). Good decomposition of `EmitCreateInstance` into named methods. Stream overloads for testability. 219 tests with solid edge case coverage.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ✅ LGTM

Found **0 issues**.

Checked against all 14 review rule categories. No violations found:
- ✅ No null-forgiving `!` operators
- ✅ No `Array.Empty<T>()` or `string.Empty` in new code
- ✅ No empty catch blocks (one `catch (IOException)` in test cleanup is properly typed)
- ✅ No `#nullable enable` (project-level nullable covers it)
- ✅ JI-style activation fully preserved (JavaInterop ctors, JNIEnv.DeleteRef cleanup)
- ✅ Proxy name collision guard preserved
- ✅ Content-dependent deterministic MVID via ComputeContentFingerprint
- ✅ Implementor/EventDispatcher trimmability heuristic properly documented
- ✅ Mono formatting style (tabs, space before parens)

👍 Clean layering of UCO wrappers, JCW generation, and JNI signature parsing on top of the #10808 proxy groundwork. 219 tests with good edge case coverage. Stream overloads for testability.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ✅ LGTM

Found **0 issues**.

Thorough review against all 14 rule categories:
- ✅ No null-forgiving `!` operators in any added line
- ✅ No `Array.Empty<T>()`, `string.Empty`, or `#nullable enable` in new code
- ✅ No empty catch blocks (test cleanup uses typed `catch (IOException)`)
- ✅ JI-style activation fully preserved from #10808 (EmitCreateInstanceViaJavaInteropNewobj, JNIEnv.DeleteRef, EmitCreateInstanceBodyWithLocals)
- ✅ Proxy name collision guard (`usedProxyNames`) preserved
- ✅ Content-dependent deterministic MVID (ComputeContentFingerprint + DeterministicMvid)
- ✅ Implementor/EventDispatcher trimmability heuristic with clear doc comment explaining the trade-off
- ✅ Forced `IgnoresAccessChecksTo("Mono.Android")` for internal JNIEnv.DeleteRef access
- ✅ Proper input validation in JniSignatureHelper (throws on malformed signatures)
- ✅ Mono formatting style throughout (tabs, space before parens)

👍 Clean layering: scanner → model → emitter pipeline. UCO wrappers (UnmanagedCallersOnly + RegisterNatives) and JCW Java source generation integrate well with the existing proxy type infrastructure. 219 tests with good coverage of edge cases (generics, nested types, JI-style activation, aliases, Implementor/EventDispatcher heuristics). Stream-based test overloads avoid flaky disk I/O.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ✅ LGTM

Found **0 issues**.

Full review against all 14 rule categories — no violations:
- ✅ No null-forgiving `!`, no `Array.Empty<T>()`, no `string.Empty`, no `#nullable enable`, no `#region`
- ✅ No empty catch blocks (test cleanup uses typed `catch (IOException)`)
- ✅ JNI boolean correctly encoded as `byte` for UCO blittability
- ✅ JI-style activation preserved (JniObjectReference conversion + JNIEnv.DeleteRef)
- ✅ Content-dependent deterministic MVID
- ✅ Proxy name collision guard
- ✅ Mono formatting (tabs, space before parens)
- ✅ 219 tests with stream-based helpers

👍 Well-structured pipeline: scanner → model → emitter. UCO wrappers, JCW Java source, and JNI signature parsing integrate cleanly with #10808's proxy groundwork.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ✅ LGTM

Found **0 issues**.

Reviewed all 17 changed files against 14 rule categories. Read full source files (not just diff) to understand invariants and call patterns.

**Checked and verified:**
- No null-forgiving `!`, no `Array.Empty<T>()` in new code, no `string.Empty`, no `#nullable enable`, no empty catch blocks
- JNI boolean correctly encoded as `byte` (blittable for `[UnmanagedCallersOnly]`)
- JI-style activation preserved from `#10808` (JniObjectReference conversion + JNIEnv.DeleteRef cleanup)
- Content-dependent deterministic MVID (ComputeContentFingerprint)
- Proxy name collision guard (usedProxyNames)
- Implementor/EventDispatcher trimmability heuristic with documented trade-off
- Input validation on all public methods; internal `Generate(type, writer)` has single call site with pre-validated args
- No resource leaks — all IDisposable properly managed

👍 Clean scanner → model → emitter pipeline. UCO wrappers, JCW Java source, and JNI signature parsing integrate well with the #10808 proxy groundwork. 219 tests with good edge-case coverage and stream-based helpers.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Comment by simonrozsival** [comment]
@jonathanpeppers this is ready for review. Adds UCO wrappers, JCW Java source generator, and JNI signature helpers on top of #10808. `GetFunctionPointer` dispatch will follow in a stacked PR.

File: src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs [hunk 8]
```diff
@@ -154,10 +180,16 @@ void EmitTypeReferences ()
 			metadata.GetOrAddString ("Java.Interop"), metadata.GetOrAddString ("JniObjectReference"));
 		_jniObjectReferenceOptionsRef = metadata.AddTypeReference (_javaInteropRef,
 			metadata.GetOrAddString ("Java.Interop"), metadata.GetOrAddString ("JniObjectReferenceOptions"));
+		_iAndroidCallableWrapperRef = metadata.AddTypeReference (_pe.MonoAndroidRef,
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply src/Microsoft.Android.Sdk.TrimmableTypeMap/Generator/TypeMapAssemblyEmitter.cs:183]: I think it's confused because `Mono.Android.dll` has a `Java.Interop` namespace inside. We can ignore this comment.

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/TypeMapModelBuilderTests.cs [hunk 9]
```diff
@@ -457,8 +457,8 @@ public void Fixture_HelperType_IsUnconditional (string javaName, string kind)
 			var entry = model.Entries.FirstOrDefault ();
 			Assert.NotNull (entry);
 			// Implementor/EventDispatcher types are treated as unconditional ACW types.
-			// Future optimization (see issue tracking Implementor trimming) may make them trimmable.
-			Assert.True (entry!.IsUnconditional, $"{kind} should be unconditional");
+			// Future optimization (see #10911) may make them trimmable.
+			Assert.True (entry.IsUnconditional, $"{kind} should be unconditional");
 		}
```
[the remark this answers is not in the document]
Comment (jonathanpeppers) [reply tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/TypeMapModelBuilderTests.cs:462]: What? No.

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs [hunk 10]
```diff
@@ -109,4 +125,21 @@ private protected static List<string> GetMemberRefNames (MetadataReader reader)
...
+		Directory.CreateDirectory (dir);
+		return dir;
+	}
+
+	private protected static void DeleteTempDir (string dir)
+	{
+		if (Directory.Exists (dir)) {
+			try {
+				Directory.Delete (dir, true);
+			} catch (IOException) {
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/FixtureTestBase.cs]: 🤖 ⚠️ **Error handling** — This catch block silently swallows `IOException`. While this is test cleanup code, the convention is to never have empty catch blocks. Consider at minimum adding a comment explaining why it's acceptable (e.g., `// Best-effort cleanup — ignore failures during temp dir deletion`).

_Rule: No empty catch blocks (Postmortem `#11`)_

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/JcwJavaSourceGeneratorTests.cs [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+		{
+			var java = GenerateFixture ("my/app/TouchHandler");
+			AssertContainsLine ("public boolean onTouch (android.view.View p0, int p1)\n", java);
+			AssertContainsLine ("public void onScroll (int p0, float p1, long p2, double p3)\n", java);
+			AssertContainsLine ("public java.lang.String getText ()\n", java);
+			AssertContainsLine ("public void setItems (java.lang.String[] p0)\n", java);
+		}
+
+	}
+}
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/JcwJavaSourceGeneratorTests.cs]: 🤖 💡 **Formatting** — Missing trailing newline at end of file. Same for `TypeMapAssemblyGeneratorTests.cs` and `TypeMapModelBuilderTests.cs`.

_Rule: Files should end with a newline_

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary

**Verdict**: ⚠️ Needs Changes

Found **3 issues**: 1 warning and 2 suggestions. CI is still pending (build `1332295`).

- ⚠️ **Error handling**: Empty catch block swallows `IOException` silently in `DeleteTempDir` (`FixtureTestBase.cs:141`)
- 💡 **Formatting**: Missing newline at end of file in `JcwJavaSourceGeneratorTests.cs`, `TypeMapAssemblyGeneratorTests.cs`, and `TypeMapModelBuilderTests.cs`
- 💡 **Testing**: `ClassDeclaration`, `AbstractType`, and `NestedType` tests use `Assert.Contains("...\n", java)` while the 5 analogous failing tests use `AssertContainsLine`. These happen to work because they assert `writer.Write($"...\n")` output (not raw strings), but consider using `AssertContainsLine` throughout for consistency

👍 Excellent test coverage — 215 tests covering all generators, model builder, scanner, and edge cases. The JNI validation in `ValidateJniName` properly prevents path traversal. The UCO/RegisterNatives emission architecture is well-documented with clear XML docs and a helpful pseudo-Java example in the class remarks. The `AssertContainsLine` helper is a clean fix for the cross-platform newline issue. Good use of `record` types for data models.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

File: tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/TypeMapAssemblyGeneratorTests.cs [hunk 12]
```diff
@@ -410,4 +410,98 @@ public void Generate_IdenticalContent_ProducesIdenticalMVIDs ()
...
+	{
+		Assert.ThrowsAny<ArgumentException> (() => JniSignatureHelper.ParseParameterTypes ("not-a-sig"));
+	}
+
+	[Fact]
+	public void ParseParameterTypes_UnterminatedSignature_ReturnsEmptyList ()
+	{
+		Assert.Empty (JniSignatureHelper.ParseParameterTypes ("("));
+	}
 }
```
Comment (simonrozsival) [inline tests/Microsoft.Android.Sdk.TrimmableTypeMap.Tests/Generator/TypeMapAssemblyGeneratorTests.cs]: 🤖 💡 **Formatting** — Missing trailing newline at end of file. Same for `TypeMapModelBuilderTests.cs`.

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary (v2)

**Verdict**: ⚠️ Needs Changes

Found **2 issues**: 1 warning and 1 suggestion. Previous review findings (empty catch block, temp dir IO) have been addressed.

- ⚠️ **CI**: Linux build is failing (`CS0246: AdbDeviceInfo`) — this appears to be a pre-existing `main` issue from the `xamarin-android-tools` submodule bump, not caused by this PR. Mac/Windows legs are still pending.
- 💡 **Formatting**: `TypeMapAssemblyGeneratorTests.cs` and `TypeMapModelBuilderTests.cs` are missing trailing newlines

👍 Clean diff — no `null!`, no empty catch blocks, no `Array.Empty`, no disk IO in tests. The `AssertContainsLine` helper properly handles cross-platform line endings. 214 tests cover generators, model builder, scanner, and edge cases. The JNI validation prevents path traversal. UCO/RegisterNatives architecture is well-documented.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by simonrozsival — COMMENTED** [review]
## 🤖 AI Review Summary (v3)

**Verdict**: ⚠️ Needs Changes

All previous findings have been addressed. **No new code issues found.**

- ⚠️ **CI**: All legs are pending (build `1332407`). The previous run's Linux failure (`CS0246: AdbDeviceInfo`) was a pre-existing `main` issue from the `xamarin-android-tools` submodule bump — not caused by this PR. Cannot confirm merge readiness until CI is green.

👍 Clean diff: no `null!`, no empty catch blocks, no `Array.Empty`, no disk IO in tests, all files end with newlines. `AssertContainsLine` properly handles cross-platform line endings. 214 tests cover generators, model builder, scanner, and edge cases.

---
_Review generated by android-reviewer from [review guidelines](https://github.com/dotnet/android/blob/main/.github/skills/android-reviewer/references/review-rules.md)._

**Review by jonathanpeppers — APPROVED** [review]
